earthfile2llb/wait_block.go

```diff
@@ -0,0 +1,190 @@
+			}
+			if !hasPlatform {
+				return fmt.Errorf("SAVE IMAGE %s was already declared (none had --platform values)", saveImage.si.DockerTag)
+			}
+			if _, found := noManifestListImgs[saveImage.si.DockerTag]; found {
+				return fmt.Errorf("cannot save image %s defined multiple times, but declared as SAVE IMAGE --no-manifest-list", saveImage.si.DockerTag)
+			}
+		}
+
+		if saveImage.si.HasPlatform {
+			// SAVE IMAGE was called with a --platform value
+			if saveImage.si.NoManifestList {
+				noManifestListImgs[saveImage.si.DockerTag] = struct{}{}
+				isMultiPlatform[saveImage.si.DockerTag] = false
+			} else {
+				isMultiPlatform[saveImage.si.DockerTag] = true // do I need to count for previsouly seen?
+			}
+		} else {
+			isMultiPlatform[saveImage.si.DockerTag] = false
+		}
+		imageWaitItems = append(imageWaitItems, saveImage)
+	}
+	if len(imageWaitItems) == 0 {
+		return nil
+	}
+
+	metadata := map[string][]byte{}
+	refs := map[string]client.Reference{}
+
+	imageSaver := saveimageutil.NewExportImageSaver(refs, metadata)
+
+	refID := 0
+	for _, item := range imageWaitItems {
+		if !item.push {
+			continue
+		}
+
+		ref, err := llbutil.StateToRef(
+			ctx, item.c.opt.GwClient, item.si.State, item.c.opt.NoCache,
+			item.c.platr, item.c.opt.CacheImports.AsMap())
+		if err != nil {
+			return errors.Wrapf(err, "failed to solve image required for %s", item.si.DockerTag)
+		}
+
+		var platformBytes []byte
+		if isMultiPlatform[item.si.DockerTag] {
+			platformBytes = []byte(item.si.Platform.String())
+			platformImgName, err := llbutil.PlatformSpecificImageName(item.si.DockerTag, item.si.Platform)
+			if err != nil {
+				return err
+			}
+
+			if item.si.CheckDuplicate && item.si.DockerTag != "" {
+				if _, found := platformImgNames[platformImgName]; found {
+					return errors.Errorf(
+						"image %s is defined multiple times for the same platform (%s)",
+						item.si.DockerTag, platformImgName)
+				}
+				platformImgNames[platformImgName] = true
+			}
+		}
+
+		_, err = imageSaver.AddPushImageEntry(ref, refID, item.si.DockerTag, true, item.si.InsecurePush, item.si.Image, platformBytes)
+		if err != nil {
+			return err
+		}
+		refID++
+	}
+
+	if len(imageWaitItems) == 0 {
+		panic("saveImagesWaitItem should never have been created with zero converters")
+	}
+	gatewayClient := imageWaitItems[0].c.opt.GwClient // could be any converter's gwClient (they should app be the same)
+
+	err := gatewayClient.Export(ctx, gwclient.ExportRequest{
+		Refs:     refs,
+		Metadata: metadata,
+	})
+	if err != nil {
+		return errors.Wrap(err, "failed to SAVE IMAGE")
+	}
+	return nil
+}
+
+func (wb *waitBlock) waitCommands(ctx context.Context) error {
+	errGroup, ctx := serrgroup.WithContext(ctx)
+
+	for _, item := range wb.items {
+		cmdItem, ok := item.(*runCommandWaitItem)
+		if !ok {
+			continue
+		}
+		errGroup.Go(func() error {
+			return cmdItem.c.forceExecution(ctx, *cmdItem.cmd, cmdItem.c.platr)
+		})
+	}
+	return errGroup.Wait()
+}
```

Comment:
Thinking about this some more... I think calling `forceExecution` on targets other than the current one can result in buggy behavior. Here's why...

For the most part, so far, we've assumed that every converter instance can manage its own execution flow and that nothing external can trigger execution unless the converter has finished processing entirely (`FinalizeStates` has been called). It feels that the wait logic in here breaks these assumptions and it's really hard to tell whether there will be things that will no longer work when that happens. And even if there aren't the execution flow becomes much harder to reason about, which could open us up in the future to bugs we cannot explain or debug.

I would do the following:

* if a target is part of a `WAIT` block in its entirety (e.g. via `BUILD +target`, `FROM +target` within a `WAIT`), I would use only `BuildAsync` + the callback for the waiting. That is the main way we wait for targets to complete today in other parts of our code (e.g. in `WITH DOCKER` code).
* if a target is part of a `WAIT` block only partially (basically the target that contains the `WAIT` itself), then we can perform `forceExecution` for the waiting (since now we are within the same converter thread, which means we have ownership over the control of the execution of the converter).

I think this also has the benefit that it keeps the control flow of the converter more manageable, and easier to reason about in our general design.